### PR TITLE
LibJS: Remove DeprecatedString usage from SourceCode

### PR DESCRIPTION
**Pull request by evansmal**
This change removes usage of DeprecatedString from the LibJS parsing code. The main changes are in Token, SourceCode, and Lexer but some other small modifications were required to meet their updated interfaces.

### CODE REVIEW

File: Userland/Libraries/LibJS/AST.cpp [hunk 1]
```diff
@@ -4962,9 +4962,9 @@ ModuleRequest::ModuleRequest(DeprecatedFlyString module_specifier_, Vector<Asser
     });
 }
 
-DeprecatedString const& SourceRange::filename() const
```
Comment (evansmal) [inline Userland/Libraries/LibJS/AST.cpp:4965]: Alternatively could store a copy of `code->filename().to_deprecated_string()` and return a reference to it.

File: Userland/Libraries/LibJS/Parser.cpp [hunk 2]
```diff
@@ -409,7 +409,7 @@ Parser::ParserState::ParserState(Lexer l, Program::Type program_type)
 }
 
 Parser::Parser(Lexer lexer, Program::Type program_type, Optional<EvalInitialState> initial_state_for_eval)
-    : m_source_code(SourceCode::create(lexer.filename(), lexer.source()))
+    : m_source_code(SourceCode::create((String::from_deprecated_string(lexer.filename()).value()), String::from_deprecated_string(lexer.source()).value()))
```
Comment (evansmal) [inline Userland/Libraries/LibJS/Parser.cpp:412]: Unsure what makes the most sense here. There's no `vm` reference available within Parser so `TRY_OR_THROW_OOM` is not easily usable.

File: Userland/Libraries/LibJS/Lexer.cpp [hunk 3]
```diff
@@ -24,7 +24,7 @@ HashMap<char, TokenType> Lexer::s_single_char_tokens;
 Lexer::Lexer(StringView source, StringView filename, size_t line_number, size_t line_column)
     : m_source(source)
     , m_current_token(TokenType::Eof, {}, {}, {}, filename, 0, 0, 0)
-    , m_filename(filename)
+    , m_filename(String::from_utf8(filename).release_value_but_fixme_should_propagate_errors())
```
Comment (evansmal) [inline Userland/Libraries/LibJS/Lexer.cpp:27]: There is likely a better way - let me know what would be idiomatic here.

File: Userland/Libraries/LibJS/Token.h [hunk 4]
```diff
@@ -181,7 +182,7 @@ class Token {
 public:
     Token() = default;
 
-    Token(TokenType type, DeprecatedString message, StringView trivia, StringView value, StringView filename, size_t line_number, size_t line_column, size_t offset)
```
Comment (davidot) [inline Userland/Libraries/LibJS/Token.h:184]: Why change this one if the underlying `m_message` is still a DeprecatedString?

Feel free to update that as well, combined with some of the changes in Lexer in its own commit.

File: Userland/Libraries/LibJS/SourceCode.cpp [hunk 5]
```diff
@@ -12,23 +12,23 @@
 
 namespace JS {
 
-NonnullRefPtr<SourceCode> SourceCode::create(DeprecatedString filename, DeprecatedString code)
+NonnullRefPtr<SourceCode> SourceCode::create(String filename, String code)
 {
-    return adopt_ref(*new SourceCode(move(filename), move(code)));
+    return adopt_ref(*new SourceCode(std::move(filename), std::move(code)));
```
Comment (davidot) [inline Userland/Libraries/LibJS/SourceCode.cpp:17]: Please leave the `move`s as is, don't add `std` (I know that annoyingly IDEs sometimes autocomplete this)

File: Userland/Libraries/LibJS/Lexer.cpp [hunk 6]
```diff
@@ -836,7 +836,7 @@ Token Lexer::next()
 
     if (m_hit_invalid_unicode.has_value()) {
         value_start = m_hit_invalid_unicode.value() - 1;
-        m_current_token = Token(TokenType::Invalid, "Invalid unicode codepoint in source",
+        m_current_token = Token(TokenType::Invalid, String::from_deprecated_string("Invalid unicode codepoint in source").release_value_but_fixme_should_propagate_errors(),
```
Comment (davidot) [inline Userland/Libraries/LibJS/Lexer.cpp:839]: Currently the three changes (this one and the two below) are about Token.
I think we should either drop these changes or update Token in a seperate commit.

**Review by davidot — CHANGES_REQUESTED** [review]
Changes look fine to me, I'm just a bit confused by the Token changes.
As I propose in the comments, I would say either split it up into two commits or drop those changes.

File: Userland/Libraries/LibJS/AST.cpp (hunk 1 above)
Comment (davidot) [reply Userland/Libraries/LibJS/AST.cpp:4965]: I think this is fine, ideally this would return a `String const&` in the future again resolving this issue.

File: Userland/Libraries/LibJS/Lexer.cpp (hunk 3 above)
Comment (davidot) [reply Userland/Libraries/LibJS/Lexer.cpp:27]: You would have to make a factory function (the `static Class::create` ones) but that would mean Parser would also need one and so on.
So for now I think this is fine.

File: Userland/Libraries/LibJS/Lexer.cpp (hunk 6 above)
Comment (trflynn89) [reply Userland/Libraries/LibJS/Lexer.cpp:839]: This one should actually be:
```c++
String::from_utf8("Invalid unicode codepoint in source"sv).release_value_but_fixme_should_propagate_errors()
```

File: Userland/Libraries/LibJS/SourceCode.cpp (hunk 5 above)
Comment (evansmal) [reply Userland/Libraries/LibJS/SourceCode.cpp:17]: My STL muscle memory must have taken over. Fixed.

File: Userland/Libraries/LibJS/Lexer.cpp (hunk 6 above)
Comment (evansmal) [reply Userland/Libraries/LibJS/Lexer.cpp]: Fixed.

File: Userland/Libraries/LibJS/Token.h (hunk 4 above)
Comment (evansmal) [reply Userland/Libraries/LibJS/Token.h:184]: I decided to update Token to use String and it is now here under a separate commit.

**Review by davidot — APPROVED** [review]
Yeah looks great now!

File: Userland/Libraries/LibJS/Parser.cpp [hunk 7]
```diff
@@ -4135,8 +4135,8 @@ void Parser::expected(char const* what)
 {
     auto message = m_state.current_token.message();
     if (message.is_empty())
-        message = DeprecatedString::formatted("Unexpected token {}. Expected {}", m_state.current_token.name(), what);
-    syntax_error(message);
+        message = String::formatted("Unexpected token {}. Expected {}", m_state.current_token.name(), what).release_value_but_fixme_should_propagate_errors();
```
Comment (linusg) [inline Userland/Libraries/LibJS/Parser.cpp:4138]: How is this related to SourceCode?

File: Userland/Libraries/LibJS/Runtime/Error.cpp [hunk 8]
```diff
@@ -54,7 +54,7 @@ ThrowCompletionOr<void> Error::install_error_cause(Value options)
 
 void Error::populate_stack()
 {
-    static auto dummy_source_range = SourceRange { .code = SourceCode::create("", ""), .start = {}, .end = {} };
+    static auto dummy_source_range = SourceRange { .code = SourceCode::create(String(), String()), .start = {}, .end = {} };
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/Error.cpp:57]: Please use `String {}` which is in line with empty string uses in other places.

File: Userland/Libraries/LibJS/Lexer.cpp [hunk 9]
```diff
@@ -893,7 +893,7 @@ Token Lexer::force_slash_as_regex()
 
     m_current_token = Token(
         token_type,
-        "",
+        String(),
```
Comment (linusg) [inline Userland/Libraries/LibJS/Lexer.cpp:896]: Likewise

File: Userland/Libraries/LibJS/Token.h [hunk 10]
```diff
@@ -8,6 +8,7 @@
 
 #include <AK/DeprecatedFlyString.h>
 #include <AK/DeprecatedString.h>
+#include <AK/String.h>
```
Comment (linusg) [inline Userland/Libraries/LibJS/Token.h:11]: Forgot to remove the include

File: Userland/Libraries/LibJS/Lexer.cpp (hunk 3 above)
Comment (linusg) [inline Userland/Libraries/LibJS/Lexer.cpp:27]: You're still mixing stuff up in commits, make a separate one for the Leyer changes - they seem totally unrelated

**Review by linusg — CHANGES_REQUESTED** [review]

File: Userland/Libraries/LibJS/Lexer.cpp (hunk 3 above)
Comment (evansmal) [reply Userland/Libraries/LibJS/Lexer.cpp:27]: Moved to a separate commit

File: Userland/Libraries/LibJS/Parser.cpp (hunk 7 above)
Comment (evansmal) [reply Userland/Libraries/LibJS/Parser.cpp]: It seems less intrusive to use ` auto message = m_state.current_token.message().to_deprecated_string();` here instead, so I've updated this change.

**Comment by evansmal** [comment]
@linusg Thanks for the comments. :smile: 

 Ready for another review!

File: Userland/Libraries/LibJS/Token.h [hunk 11]
```diff
@@ -181,7 +181,7 @@ class Token {
 public:
     Token() = default;
 
-    Token(TokenType type, DeprecatedString message, StringView trivia, StringView value, StringView filename, size_t line_number, size_t line_column, size_t offset)
+    Token(TokenType type, String const& message, StringView trivia, StringView value, StringView filename, size_t line_number, size_t line_column, size_t offset)
```
Comment (linusg) [inline Userland/Libraries/LibJS/Token.h:184]: This should still take the string by value (and then `move()` it into `m_message`, which is currently missing)

**Review by linusg — CHANGES_REQUESTED** [review]
Please make sure all commits compile on their own, currently they don't.

File: Userland/Libraries/LibJS/Token.h (hunk 11 above)
Comment (evansmal) [reply Userland/Libraries/LibJS/Token.h]: Fixed

**Comment by evansmal** [comment]
> Please make sure all commits compile on their own, currently they don't.

All commits should now compile on their own.

**Review by linusg — APPROVED** [review]
Thanks!
